### ISSUE TITLE
Improve ollama's Output Speed

### ISSUE THREAD

**Issue by System233** [ollama/ollama#5416]
I have noticed that ollama always outputs content at a fixed speed, and most of the time, the GPU is not fully utilized(0% load), indicating that the bottleneck in generating content is not in the GPU.

At the same time, I have observed that during content output, one CPU core consistently has a high load(70% load), suggesting that the bottleneck might be in the CPU decoding process.

Therefore, could you consider implementing multi-threaded decoding to leverage the multi-core capabilities of the CPU to accelerate the output speed?

I have already searched Google for discussions on this issue but did not find anything relevant.

If my analysis is incorrect or the issue cannot be resolved, you can directly close this issue. Thank you.

**Comment by KastanDay** [comment]
You have 0% GPU load for the entire generation process? That seems like your GPU isn't being utilized. Is the GPU memory filling up? Pro tip: using `nvitop` or `nvtop` to monitor.

**Comment by rick-github** [comment]
ollama use https://github.com/ggerganov/llama.cpp for inference, so it might be better asked there.

**Comment by System233** [comment]
> You have 0% GPU load for the entire generation process? That seems like your GPU isn't being utilized. Is the GPU memory filling up? Pro tip: using or to monitor.`nvitop``nvtop`

No, what I mean is that the GPU completes the generation task very quickly, with a very short period of load time. The remaining time is spent on the CPU, which takes a long time to output the content word by word. In other words, the decoding speed of the CPU is much slower than the generation speed of the GPU.

I observed this phenomenon in the Task Manager.

**Comment by System233** [comment]
> ollama use https://github.com/ggerganov/llama.cpp for inference, so it might be better asked there.

Thanks

**Comment by KastanDay** [comment]
Very weird. I personally haven't seen this, maybe because I mostly use streaming. 

Hopefully others in the community can help you. Good luck!
